src/providers/azure.rs

```diff
@@ -0,0 +1,503 @@
+
+impl super::MachineSetup for Setup {
+    type Region = Region;
+
+    fn region(&self) -> Self::Region {
+        self.region
+    }
+}
+
+impl Setup {
+    pub fn region(self, r: Region) -> Self {
+        Self { region: r, ..self }
+    }
+
+    /// To view the available sizes in the relevant region, use:
+    /// `az vm list-sizes -l <region_name>`.
+    pub fn instance_type(self, inst_type: String) -> Result<Self, Error> {
+        if azcmd::available_instances_in_region(self.region)?
+            .iter()
+            .any(|x| x == &inst_type)
+        {
+            Ok(Self {
+                instance_type: inst_type,
+                ..self
+            })
+        } else {
+            Err(format_err!(
+                "{} not valid instance type in {:?}",
+                inst_type,
+                self.region
+            ))
+        }
+    }
+
+    /// The provided callback (`setup`) is called once for every spawned instances of this type with a handle
+    /// to the target machine. Use [`Machine::ssh`](crate::Machine::ssh) to issue
+    /// commands on the host in question.
+    pub fn setup(
+        mut self,
+        setup: impl Fn(&mut ssh::Session, &slog::Logger) -> Result<(), Error> + Send + Sync + 'static,
+    ) -> Self {
+        self.setup_fn = Some(Arc::new(setup));
+        self
+    }
+}
+
+/// This implementation relies on the [Azure
+/// CLI](https://docs.microsoft.com/en-us/cli/azure/install-azure-cli?view=azure-cli-latest). 
+///
+/// It also assumes you have previously run `az login` to authenticate with Microsoft.
+#[derive(Default)]
+pub struct AzureLauncher {
+    regions: HashMap<Region, AzureRegion>,
+}
+
+impl super::Launcher for AzureLauncher {
+    type MachineDescriptor = Setup;
+
+    fn launch(&mut self, l: super::LaunchDescriptor<Self::MachineDescriptor>) -> Result<(), Error> {
+        let region = l.region;
+        let mut az_region = AzureRegion::new(&l.region.to_string(), l.log.clone())?;
+        az_region.launch(l)?;
+        self.regions.insert(region, az_region);
+        Ok(())
+    }
+
+    fn connect_all<'l>(&'l self) -> Result<HashMap<String, crate::Machine<'l>>, Error> {
+        collect!(self.regions)
+    }
+}
+
+impl std::ops::Drop for AzureLauncher {
+    fn drop(&mut self) { }
+}
+
+struct Descriptor {
+    name: String,
+    ip: String,
+}
+
+/// Region-specific connection to Azure. Each instance of this type creates one Azure
+/// "resource group" and deletes the group on drop.
+///
+/// This implementation relies on the [Azure
+/// CLI](https://docs.microsoft.com/en-us/cli/azure/install-azure-cli?view=azure-cli-latest). 
+///
+/// It also assumes you have previously run `az login` to authenticate with Microsoft.
+#[derive(Default)]
+pub struct AzureRegion {
+    pub log: Option<slog::Logger>,
+    pub region: Region,
+    resource_group_name: String,
+    machines: Vec<Descriptor>,
+}
+
+impl AzureRegion {
+    pub fn new(region: &str, log: slog::Logger) -> Result<Self, Error> {
```

Comment:
Why does this take a `&str` for region, not a `Region`?

Reply:
oops